drivers: platform: xilinx: Updated i2c driver

scuciurean opened the pull request:
Removed platform_drivers.h dependency and updated compiler
switches to support both PS and PL implementation of a i2c
controller.
The driver is now updated to use the platform agnostic extra
features passed in the i2c descriptor.
Changed 'stop_bit' param to 'options'.

Signed-off-by: Sergiu Cuciurean sergiu.cuciurean@analog.com

dbogdan commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -57,65 +67,98 @@
 #endif
+	/* Intended fallthrough */
+	default:
+		goto not_defined;
```
Shouldn't you goto "error" instead?

scuciurean replied:
Got updated.

adrimbarean commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -57,65 +67,98 @@
+		Xil_ExceptionInit();
+		Xil_ExceptionRegisterHandler(XIL_EXCEPTION_ID_IRQ_INT,
+					     XIic_InterruptHandler,
+					     &xdesc->instance);
```
This part is a bit tricky. If the core should be the only one to generate interrupts in the system it is OK to connect it directly to the ALU interrupt handler. But if there are more cores need to generate interrupts there would need to be an interrupt controller that will be connected to the ALU handler and all the other ones are connected to the interrupt controller. For example when using the I2C core and a timer, or an UART core.

Maybe the driver could check if there is an interrupt controller available, and if there is connect to it instead, and if there is not then do what it does now. Maybe this check can be done with and XPAR_NUM_OF_INSTANCES like the availability of the I2C cores?

adrimbarean commented on drivers/platform/xilinx/i2c_extra.h:
```diff
@@ -0,0 +1,69 @@
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+typedef enum xil_i2c_type {
```
Should we still be using `typedef`s?

scuciurean replied:
@dbogdan Should we?

dbogdan replied:
We should comply with the Linux kernel coding style: https://github.com/analogdevicesinc/linux/blob/master/Documentation/process/coding-style.rst#5-typedefs

scuciurean commented:
Changelog:

1. Fixed _goto error_ statement in i2c_init to free the memory properly.

mirceacaprioru commented:
there is a codacy complaint.

adrimbarean commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -152,16 +214,55 @@ int32_t i2c_remove(struct i2c_desc *desc)
-			 bytes_number, stop_bit ? XIIC_STOP : XIIC_REPEATED_START);
-#else
-	return SUCCESS;
+		ret = XIicPs_SetOptions(xdesc->instance,
```
`XIicPs_SetOptions` is from the PS driver not PL. The PL equivalent would be `XIic_SetOptions` and it returns void, so the return value cannot be tested here. Check for the `i2c_write` function too.

adrimbarean commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -57,65 +67,98 @@
+		Xil_ExceptionInit();
+		Xil_ExceptionRegisterHandler(XIL_EXCEPTION_ID_IRQ_INT,
+					     XIic_InterruptHandler,
+					     &xdesc->instance);
```
Also the `instance` member of `xdesc` is already a pointer (void pointer) to the struct we need. Giving it by reference here would make it a double pointer. Reference needs to be removed and maybe cast is needed.

adrimbarean commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -57,65 +67,98 @@
+		Xil_ExceptionRegisterHandler(XIL_EXCEPTION_ID_IRQ_INT,
+					     XIic_InterruptHandler,
+					     &xdesc->instance);
+		ret = XIic_Start(&xdesc->instance);
```
`instance` member of `xdesc` is already a pointer (void pointer) to the struct we need. Giving it by reference here would make it a double pointer. Reference needs to be removed and maybe cast is needed.

adrimbarean commented on drivers/platform/xilinx/i2c.c:
```diff
@@ -57,65 +67,98 @@
-	xil_dev->id = xil_param->id;
-	dev->max_speed_hz = param->max_speed_hz;
-	dev->slave_address = param->slave_address;
+		ret = XIic_SetAddress(&xdesc->instance,
```
Double pointer here too, as above.

scuciurean commented:
Changelog -> V3:

1. Removed the interrupt assignment to the main IRQ.
2. Updated how the i2c descriptor is passed in several functions.
3. Removed status check for functions that don't return a status.

dbogdan commented:
The README changes have a different reason: I2C is not used in those projects, so we should have a different commit for it.
Good guidelines regarding this can be found here: https://github.com/analogdevicesinc/linux/blob/master/Documentation/process/submitting-patches.rst

dbogdan commented:
Don't forget to also update the body of the explanation line wrapping - if I'm not wrong, it is now set to 60 columns for your commit.

scuciurean commented:
Changelog -> V4:

1. Moved README update to a different commit
